mm2src/coins/solana/solana_common.rs

```diff
@@ -0,0 +1,59 @@
+use crate::solana::{SolanaAsyncCommonOps, SolanaCommonOps};
+use crate::{MarketCoinOps, WithdrawError, WithdrawRequest};
+use bigdecimal::ToPrimitive;
+use common::mm_error::MmError;
+use common::mm_number::BigDecimal;
+use futures::compat::Future01CompatExt;
+use solana_sdk::hash::Hash;
+use solana_sdk::native_token::LAMPORTS_PER_SOL;
+
+pub fn lamports_to_sol(lamports: u64) -> BigDecimal { BigDecimal::from(lamports) / BigDecimal::from(LAMPORTS_PER_SOL) }
+
+pub fn sol_to_lamports(sol: &BigDecimal) -> Option<u64> { (sol * BigDecimal::from(LAMPORTS_PER_SOL)).to_u64() }
+
+pub fn ui_amount_to_amount(ui_amount: BigDecimal, decimals: u8) -> Option<u64> {
+    (ui_amount * BigDecimal::from(10_u64.pow(decimals as u32))).to_u64()
+}
+
+pub fn amount_to_ui_amount(amount: u64, decimals: u8) -> BigDecimal {
+    BigDecimal::from(amount) / BigDecimal::from(10_u64.pow(decimals as u32))
+}
+
+pub async fn check_sufficient_balance<T>(
+    coin: &T,
+    req: &WithdrawRequest,
+) -> Result<(BigDecimal, BigDecimal), MmError<WithdrawError>>
+where
+    T: SolanaCommonOps + SolanaAsyncCommonOps + MarketCoinOps,
+{
+    let my_balance = coin.my_balance().compat().await?.spendable;
+    let to_send = if req.max {
+        my_balance.clone()
+    } else {
+        req.amount.clone()
+    };
+    if to_send > my_balance {
+        return MmError::err(WithdrawError::NotSufficientBalance {
+            coin: coin.ticker().to_string(),
+            available: my_balance.clone(),
+            required: &to_send - &my_balance,
+        });
+    }
+    Ok((to_send, my_balance))
+}
+
+pub async fn check_amount_too_low<T>(coin: &T) -> Result<(BigDecimal, Hash), MmError<WithdrawError>>
```

Comment:
> The naming is highly confusing - it seems that it should have an `amount` argument, but you check that coin's balance is above a specific threshold. And it also requests recent block hash for whatever reason. `WithdrawError` also seems excessive because it can return only a single variant of it. Please use an appropriate name and error type here and avoid mixing up such stuff in general.

The function doesn't exist anymore since the fees calculation for Solana changed, they have different behaviour for SPL/Solana